frontend/middleware.ts

```diff
@@ -0,0 +1,16 @@
+import { NextResponse } from 'next/server';
+import type { NextRequest } from 'next/server';
+import { PROD_HOSTNAME } from '@config/constants';
+import { isNonNullable } from '@ifixit/helpers';
+
+type Header = [string, string] | null;
+
+export function middleware(request: NextRequest) {
+   const robotTag: Header =
+      request.nextUrl.hostname === PROD_HOSTNAME
+         ? ['X-Robots-Tag', 'noindex, nofollow']
+         : null;
+
+   const headers = [robotTag].filter(isNonNullable);
+   return NextResponse.next({ headers });
```

Comment:
This seems a little overcomplicated to me with the filter, especially since this is going to run on every single request. This way brings it down to a single conditional with no additional function calls, and in the prod case, it should be equivalent to calling the function with no params. Also, I flipped the conditional since that way it's written seems like it would noindex every page on prod *only*.
```suggestion
   const options =
      request.nextUrl.hostname !== PROD_HOSTNAME
         ? { headers: [['X-Robots-Tag', 'noindex, nofollow']] }
         : undefined;

   return NextResponse.next(options);
```
I couldn't find any docs explaining what NextResponse.next() takes, and it seemed all the examples took nothing. I saw that some next.js tests included a form like this, but that seems weird. Did you find any docs explaining what NextResponse.next takes or what?

Reply:
I chose to implement it with `.filter()` in case we wanted to add other headers on different conditions in the future. The `isNonNullable` callback was to appease typescript (given an anonymous function, typescript still thought the result array could have null values). What you have will work too, I don't have a strong opinion.

Here are the types we're working with. I found them with Ctrl + click in VSCode 🚀 . I think they're also on GH [here](https://github.com/vercel/next.js/blob/canary/packages/next/server/web/spec-extension/response.ts).

<details>

```ts

// node_modules/next/dist/server/web/spec-extension/response.d.ts

export declare class NextResponse extends Response {
    [INTERNALS]: {
        cookies: NextCookies;
        url?: NextURL;
    };
    constructor(body?: BodyInit | null, init?: ResponseInit);
    get cookies(): NextCookies;
    static json(body: any, init?: ResponseInit): NextResponse;
    static redirect(url: string | NextURL | URL, init?: number | ResponseInit): NextResponse;
    static rewrite(destination: string | NextURL | URL, init?: ResponseInit): NextResponse;
    static next(init?: ResponseInit): NextResponse;
}
interface ResponseInit extends globalThis.ResponseInit {
    nextConfig?: {
        basePath?: string;
        i18n?: I18NConfig;
        trailingSlash?: boolean;
    };
    url?: string;
}

// node_modules/typescript/lib/lib.dom.d.ts

interface globalThis.ResponseInit {
    headers?: HeadersInit;
    status?: number;
    statusText?: string;
}

/* snip */

type HeadersInit = [string, string][] | Record<string, string> | Headers;

declare var Headers: {
    prototype: Headers;
    new(init?: HeadersInit): Headers;
};
```

</details>

Reply:
Ah cool. Thank goodness for inline docs. I think we should do this the simpler way, and if we decide we need more complicated header middleware in the future, we can always add it then.